### PR TITLE
Start using final CPython 3.11 and dev 3.12 in CI

### PR DESCRIPTION
**Pull request by illia-v**

### CODE REVIEW

**Review by sethmlarson — APPROVED** [review]
Thank you, Illia! Hope you've been doing well :)

**Comment by sethmlarson** [comment]
Also, once this is merged @illia-v can you backport to `1.26.x`?

**Comment by illia-v** [comment]
> Also, once this is merged @illia-v can you backport to `1.26.x`?

Yes, I'll backport it.

---

Tests on latest CPython on macOS and on dev versions tend to time out as they used to do previously. I hope there will be less timeouts after dropping support for pyOpenSSL and SecureTransport.

**Comment by sethmlarson** [comment]
Hmm, tests timed out again. Makes me want to disable pyOpenSSL tests on Python 3.12 since it's [not out for another year](https://peps.python.org/pep-0693/) and pyOpenSSL support will be removed before then. Could you try that @illia-v?

**Comment by illia-v** [comment]
macOS 3.11 seems to time out because of `test.test_wait.test_eintr_zero_timeout` always 😕

**Comment by sethmlarson** [comment]
@illia-v That one might be worth diving deeper into, could be a bug in 3.11?

**Comment by illia-v** [comment]
> Hmm, tests timed out again. Makes me want to disable pyOpenSSL tests on Python 3.12 since it's [not out for another year](https://peps.python.org/pep-0693/) and pyOpenSSL support will be removed before then. Could you try that @illia-v?

I fixed the texts by removing usage of `ssl.wrap_socket` dropped in CPython 3.12.

It is a bit annoying that tests just hang without displaying any error if something fails in a function passed to `dummyserver.testcase.SocketDummyServerTestCase._start_server` 🙁

**Comment by illia-v** [comment]
> @illia-v That one might be worth diving deeper into, could be a bug in 3.11?

Maybe.

FYI, `test_eintr_zero_timeout[select_wait_for_socket]` (a variant of the test that does not use `select.poll`) passes successfully.
https://github.com/illia-v/urllib3/actions/runs/3380713141/jobs/5613807022

Also, all three variants of similar `test_eintr` and `test_eintr_infinite_timeout` pass successfully.
https://github.com/illia-v/urllib3/actions/runs/3380713141/jobs/5613807022

**Comment by Ousret** [comment]
Could you revert the changes about 3.12 compatibility? I don't think that testing 3.12 should pass as mandatory for now.
I have a dedicated PR on that subject. Thank you.

**Comment by sethmlarson** [comment]
@Ousret Could you explain your comment a bit more, I don't quite understand what you want to happen?

**Comment by Ousret** [comment]
Of course, 
I could be mistaken but,
- It would be more traceable to concentrate the fix on the branch that I work on, targeting 1.26x, and then make a customized port (simplified) to the main branch. 
- Solving the breakage [627c4ee](https://github.com/urllib3/urllib3/pull/2760/commits/627c4ee08a553dcca6a1047abe9749bf73eb6493) with Python 3.12 should be addressed separately. As "3.12-dev" is marked experimental, failures are not obstacles to passing the ci.
- On another note, I thought it would be interesting to further take the time to assess the impact of the forthcoming Python release to urllib3.
Hope that clarifies.

**Comment by illia-v** [comment]
@Ousret feel free to cherry-pick or copy changes related to `ssl.wrap_socket` and enabling Python 3.12 in CI to your PR.

This PR is blocked by a test failing on CPython 3.11 on macOS. I do not have a Mac to debug it right now. If nobody has an idea about the failure, I may set up a VM to debug the error in a few days.

**Comment by illia-v** [comment]
@Ousret note that testing on 3.12 has recently been blocked by this problem https://github.com/pypa/setuptools/issues/3661

**Comment by Ousret** [comment]
@illia-v I will enable 3.12-dev testing in my branch then. And apply a similar fix for ssl.wrap_socket removal while still supporting older python versions.
Duly noted the recent breakage.

**Comment by Ousret** [comment]
Regarding the CI, what I can immediately infer is:

```python
    @notWindows
    def test_ignore_broken_pipe_errors(self, monkeypatch):
        # On Windows an aborted connection raises an error on
        # attempts to read data out of a socket that's been closed.
        sock_shut = Event()
        orig_connect = HTTPConnection.connect
        # a buffer that will cause two sendall calls
        buf = "a" * 1024 * 1024 * 4
```

And the actual failure is:

```
FAILED test/with_dummyserver/test_socketlevel.py::TestBrokenPipe::test_ignore_broken_pipe_errors - urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))
```

Maybe MacOs aligned themselves on Windows for this particular case.

**Comment by illia-v** [comment]
I've made CI run on 3.12 on macOS and Windows in addition to Ubuntu to discover problems similar to the error on 3.11 on macOS earlier in the future.

**Comment by illia-v** [comment]
I realized how to fix the disutils-related error.
A new issue is a segmentation fault introduced in CPython 3.12.0a3 likely https://github.com/python/cpython/issues/100126 🤯

**Comment by sethmlarson** [comment]
@illia-v Nice find! Good to have an issue we can track to know when to upgrade.

**Review by sethmlarson — DISMISSED** [review]
LGTM, thank you @illia-v for all the work on this! :shipit:

**Comment by pquentin** [comment]
It's great to see that 3.12 alpha 4 fixed all the issues we were seeing.

**Comment by pquentin** [comment]
The checks are in weird inconsistent state, trying to close/reopen to see if this helps.

**Review by pquentin — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (pquentin) [reply .github/workflows/ci.yml]: (I removed this, and implemented https://github.com/urllib3/urllib3/pull/2887 too.)

**Comment by pquentin** [comment]
This is looking great. The only concern I have is that the macOS job is taking around 30 minutes, so it's going to time out half of the time, which I think will prevent us from getting coverage?

**Comment by illia-v** [comment]
@pquentin I [tried](https://github.com/illia-v/urllib3/commit/e0d74acfcec64197e2f066280d092641e6bf1371) running tests in parallel using pytest-xdist. They ran faster, but some flaky (probably) tests failed in more jobs.
https://github.com/illia-v/urllib3/actions/runs/4044787855

**Pull request merged**
